// next-sanity/dist/_chunks-cjs/NextStudioNoScript.cjs
"use strict";
var jsxRuntime = require("react/jsx-runtime"), react = require("react");
const style = {
  height: "100vh",
  maxHeight: "100dvh",
  overscrollBehavior: "none",
  WebkitFontSmoothing: "antialiased",
  overflow: "auto"
}, NextStudioLayoutComponent = ({ children }) => /* @__PURE__ */ jsxRuntime.jsx("div", { id: "sanity", "data-ui": "NextStudioLayout", style, children }), NextStudioLayout = react.memo(NextStudioLayoutComponent), styles = {
  outer: {
    position: "absolute",
    top: 0,
    right: 0,
    left: 0,
    bottom: 0,
    background: "#fff",
    zIndex: 1
  },
  inner: {
    position: "absolute",
    top: "50%",
    left: "50%",
    transform: "translate(-50%, -50%)",
    textAlign: "center",
    fontFamily: "helvetica, arial, sans-serif"
  }
}, NextStudioNoScript = () => /* @__PURE__ */ jsxRuntime.jsx("noscript", { children: /* @__PURE__ */ jsxRuntime.jsx("div", { style: styles.outer, children: /* @__PURE__ */ jsxRuntime.jsxs("div", { style: styles.inner, children: [
  /* @__PURE__ */ jsxRuntime.jsx("h1", { children: "JavaScript disabled" }),
  /* @__PURE__ */ jsxRuntime.jsxs("p", { children: [
    "Please ",
    /* @__PURE__ */ jsxRuntime.jsx("a", { href: "https://www.enable-javascript.com/", children: "enable JavaScript" }),
    " in your browser and reload the page to proceed."
  ] })
] }) }) });
exports.NextStudioLayout = NextStudioLayout;
exports.NextStudioNoScript = NextStudioNoScript;
